### PR TITLE
CART-974: Make self_test optionally daos_agent aware

### PR DESCRIPTION
**Pull request by mallove79**

### CODE REVIEW

File: src/tests/ftest/network/cart_self_test.py [hunk 1]
```diff
@@ -78,6 +79,14 @@ def setUp(self):
             self.server_managers[0].get_config_value("provider")
         self.cart_env["OFI_INTERFACE"] = \
             self.server_managers[0].get_config_value("fabric_iface")
+
+        self.cart_env["DEFAULT_DAOS_AGENT_DRPC_SOCK"] = \
+            "/var/run/daos_agent/daos_agent.sock"
```
Comment (frostedcmos) [inline src/tests/ftest/network/cart_self_test.py:84]: i dont think this one is envariable

File: src/include/daos/debug.h [hunk 2]
```diff
@@ -10,7 +10,9 @@
 #include <stdio.h>
 #include <daos_errno.h>
 
+#ifndef DD_FAC
 #define DD_FAC(name)	daos_##name##_logfac
+#endif /* !DD_FAC */
```
Comment (frostedcmos) [inline src/include/daos/debug.h:15]: why is this needed now?

File: src/utils/self_test/self_test.c [hunk 3]
```diff
@@ -1686,10 +1716,11 @@ int main(int argc, char *argv[])
 			{"randomize-endpoints", no_argument, 0, 'q'},
 			{"path", required_argument, 0, 'p'},
 			{"nopmix", no_argument, 0, 'n'},
+			{"use-daos-agent-env", no_argument, 0, 'u'},
 			{0, 0, 0, 0}
 		};
 
-		c = getopt_long(argc, argv, "g:m:e:s:r:i:a:btnqp:",
+		c = getopt_long(argc, argv, "g:m:e:s:r:i:a:btnqp:u:",
```
Comment (frostedcmos) [inline src/utils/self_test/self_test.c:1723]: should this not be just  'u' without ':'? ':' usually means there needs to be parameter after -u

File: src/include/daos/debug.h (hunk 2 above)
Comment (mallove79) [reply src/include/daos/debug.h:15]: It looks like there's a conflict between subsystems in the logging framework, e.g., macros for `st` (self_test) and  `mgmt` ?  In other words, the `ifndef` is a workaround for this error:

```
+ gcc -o build/dev/gcc/src/utils/self_test/self_test.o ... src/utils/self_test/self_test.c
In file included from src/include/daos/common.h:28:0,
                 from src/include/daos/drpc.h:11,
                 from src/include/daos/agent.h:10,
                 from src/utils/self_test/self_test.c:19:
src/include/daos/debug.h:13:0: error: "DD_FAC" redefined [-Werror]
 #define DD_FAC(name) daos_##name##_logfac
 ^
In file included from src/cart/crt_internal.h:14:0,
                 from src/cart/tests_common.h:18,
                 from src/utils/self_test/self_test.c:16:
src/cart/crt_debug.h:11:0: note: this is the location of the previous definition
 #define DD_FAC(name) crt_##name##_logfac
 ^
```

File: src/utils/self_test/self_test.c (hunk 3 above)
Comment (mallove79) [reply src/utils/self_test/self_test.c]: Correct.  `-u` takes no arguments.  fb4a83b4bb1b043ba7ea22d0fb0839a480c274e1

**Comment by mallove79** [comment]
There is no functional test for `--use-daos-agent-env` option, but this manual test demonstrates when the option is in use without the `OFI_*` and `CRT_*` env vars -- `self_test` successfully attaches to the group.

```diff
diff --git a/src/tests/ftest/network/cart_self_test.py b/src/tests/ftest/network/cart_self_test.py
index 729ede83a..ad03db2b9 100755
--- a/src/tests/ftest/network/cart_self_test.py
+++ b/src/tests/ftest/network/cart_self_test.py
@@ -73,12 +73,12 @@ class CartSelfTest(TestWithServers):
             "Error updating daos_server 'crt_ctx_share_addr' config setting")

         # Setup additional environment variables for the server orterun command
-        self.cart_env["CRT_CTX_SHARE_ADDR"] = str(share_addr)
+        # self.cart_env["CRT_CTX_SHARE_ADDR"] = str(share_addr)
         self.cart_env["CRT_CTX_NUM"] = "8"
-        self.cart_env["CRT_PHY_ADDR_STR"] = \
-            self.server_managers[0].get_config_value("provider")
-        self.cart_env["OFI_INTERFACE"] = \
-            self.server_managers[0].get_config_value("fabric_iface")
+        # self.cart_env["CRT_PHY_ADDR_STR"] = \
+        #   self.server_managers[0].get_config_value("provider")
+        # self.cart_env["OFI_INTERFACE"] = \
+        #     self.server_managers[0].get_config_value("fabric_iface")
         self.cart_env["DAOS_AGENT_DRPC_DIR"] = "/var/run/daos_agent/"

         self.server_managers[0].manager.assign_environment(self.cart_env, True)
```

[Here](https://gist.github.com/mallove79/fe8936f445c1b68f123e87eae04a60ed) is a run of the test logs.

File: src/tests/ftest/network/cart_self_test.py (hunk 1 above)
Comment (mallove79) [reply src/tests/ftest/network/cart_self_test.py]: 3c0a08a9e8a501e02612f38e9f4fb85456daf917

File: src/tests/ftest/network/cart_self_test.py [hunk 4]
```diff
@@ -78,6 +79,14 @@ def setUp(self):
...
+        self.cart_env["DAOS_AGENT_DRPC_DIR"] = \
+            "/var/run/daos_agent/"
+        self.cart_env["DAOS_AGENT_DRPC_SOCK_NAME"] = \
+            "daos_agent.sock"
```
[the remark this answers is not in the document]
Comment (mallove79) [reply src/tests/ftest/network/cart_self_test.py]: 3c0a08a9e8a501e02612f38e9f4fb85456daf917

File: src/utils/self_test/SConscript [hunk 5]
```diff
@@ -13,12 +13,13 @@ SELF_TEST = 'self_test.c'
...
-    Import('env', 'prereqs', 'cart_lib', 'gurt_lib', 'cart_utils_objs')
+    Import('env', 'prereqs', 'cart_lib', 'gurt_lib', 'cart_utils_objs',
+           'dc_mgmt_tgts')
 
     tenv = env.Clone(LIBS=[])
 
-    tenv.AppendUnique(LIBS=['cart', 'gurt', 'pthread', 'm'])
-    prereqs.require(tenv, 'mercury')
+    tenv.AppendUnique(LIBS=['cart', 'gurt', 'pthread', 'm', 'daos'])
+    prereqs.require(tenv, 'mercury', 'protobufc')
```
[the remark this answers is not in the document]
Comment (mallove79) [reply src/utils/self_test/SConscript:22]: `prereqs` was not added in this PR, though it seems pylint may not account for how we're [exporting `prereqs` in SConstruct](https://github.com/daos-stack/daos/blob/master/SConstruct#L383) (see [SCons Import/Export docs](https://scons.org/doc/2.0.1/HTML/scons-user.html#AEN3295)).  Can we pylint disable this message?

**Comment by mallove79** [comment]
[Tests pass](  https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-6219/8/
), except for `Simple.DRAIN[1-9]` and `DaosCore {drain, oid_alloc, rebuild}`, which occurred on master this weekend:

   [https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/5041/#showFailuresLink](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/5041/#showFailuresLink)

Not sure about this `OSAOfflineReintegration` failure, but it doesn't appear related to `self_test` or [DAOS-7948](https://jira.hpdd.intel.com/browse/DAOS-7948):
```
  stderr: b"ERROR: : check object 1155474879844188161.32 failed: 
    DER_IO(-2001): 'I/O error'\nERROR: daos: failed to check container 
     008b5830-1b34-4560-a02b-31213fe7619b: DER_IO(-2001): I/O error\n"
```

**Comment by mallove79** [comment]
Validation test is [the same as for self_test](https://github.com/daos-stack/daos/pull/6219/#issuecomment-883606314): Unset critical env vars (see below) and ensure command runs successfully.
```diff
 78 --- a/src/tests/ftest/util/apricot/apricot/test.py
 79 +++ b/src/tests/ftest/util/apricot/apricot/test.py
 88 @@ -734,6 +735,23 @@ class TestWithServers(TestWithoutServers):
 89                  cp_command = "sudo cp {} {}".format(attachinfo_file, ".")
 90                  run_command(cp_command, verbose=True, raise_exception=False)
 91                  cart_ctl.group_name.value = manager.get_config_value("name")
 92 +
 93 +                ########################################################################
 94 +                # VALIDATION TEST
 95 +                del os.environ['OFI_INTERFACE']
 96 +                del os.environ['CRT_CTX_SHARE_ADDR']
 97 +                ########################################################################
 98 +
 99                  cart_ctl.run()
100          else:
101              self.log.info(
```

File: src/utils/ctl/cart_ctl.c [hunk 6]
```diff
@@ -701,6 +716,7 @@ ctl_init()
 	D_ASSERTF(rc == 0, "crt_finalize() failed. rc: %d\n", rc);
 
 	d_log_fini();
+	dc_mgmt_fini();
```
Comment (frostedcmos) [inline src/utils/ctl/cart_ctl.c:719]: this should be   if (ctl_gdata.cg_use_daos_agent_env) {  dc_mgmt_fini(); }

also should be moved before d_log_fini()

File: src/utils/ctl/cart_ctl.c [hunk 7]
```diff
@@ -263,6 +267,8 @@ print_usage_msg(const char *msg)
 	printf("\tdon't perform 'wait for ranks' sync\n");
 	printf("-m 'log_message'\n");
 	printf("\tSpecify log message to be sent to remote server\n");
+	printf("--use-daos-agent-env\n");
+        printf("\tSet OFI and CRT_* vars through daos_agent\n");
```
[the remark this answers is not in the document]
Comment (frostedcmos) [reply src/utils/ctl/cart_ctl.c:271]: style issue needs to be fixed

File: src/utils/self_test/self_test.c [hunk 8]
```diff
@@ -1908,6 +1943,7 @@ int main(int argc, char *argv[])
 	if (all_params != NULL)
 		D_FREE(all_params);
 	d_log_fini();
+	dc_mgmt_fini();
```
Comment (frostedcmos) [inline src/utils/self_test/self_test.c:1946]: similarly here should do it only   if (use_daos_agent_vars) 

Also this should be done before d_log_fini() in case any of dc_mgmt_* code logs messages

**Review by frostedcmos — CHANGES_REQUESTED** [review]

File: src/utils/self_test/self_test.c (hunk 8 above)
Comment (mallove79) [reply src/utils/self_test/self_test.c]: 38f8e41f157ae7b000b47dd0bd1eba0641f39a35

File: src/utils/ctl/cart_ctl.c (hunk 6 above)
Comment (mallove79) [reply src/utils/ctl/cart_ctl.c]: 38f8e41f157ae7b000b47dd0bd1eba0641f39a35

File: src/utils/ctl/cart_ctl.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mallove79) [reply src/utils/ctl/cart_ctl.c]: 43fe7d9e15e15bf4a8cb6c831dd916cadb84f7bd

File: src/include/daos/debug.h (hunk 2 above)
Comment (jolivier23) [reply src/include/daos/debug.h:15]: we should probably unify all the debug stuff between cart/gurt/daos

**Review by jolivier23 — DISMISSED** [review]

**Review by frostedcmos — CHANGES_REQUESTED** [review]
As we've discussed in test weekly meeting, this needs to be changed to populate group information from daos agent response, instead of still relying on dumped attach info file.

File: src/tests/ftest/network/cart_self_test.py [hunk 9]
```diff
@@ -46,6 +46,7 @@ def __init__(self, path=""):
 
             self.repetitions = FormattedParameter("--repetitions {0}")
             self.attach_info = FormattedParameter("--path {0}")
```
Comment (sylviachanoiyee) [inline src/tests/ftest/network/cart_self_test.py:48]: I think this can be removed if it is daos agent aware

File: src/tests/ftest/network/cart_self_test.py [hunk 10]
```diff
@@ -78,6 +79,8 @@ def setUp(self):
             self.server_managers[0].get_config_value("provider")
         self.cart_env["OFI_INTERFACE"] = \
             self.server_managers[0].get_config_value("fabric_iface")
+        self.cart_env["DAOS_AGENT_DRPC_DIR"] = "/var/run/daos_agent/"
+
         self.server_managers[0].manager.assign_environment(self.cart_env, True)
 
         # Start the daos server
```
Comment (sylviachanoiyee) [inline src/tests/ftest/network/cart_self_test.py:86]: same here. we dont need to dump attach info line 89-93

File: src/tests/ftest/network/cart_self_test.py [hunk 11]
```diff
@@ -46,6 +46,7 @@ def __init__(self, path=""):
 
             self.repetitions = FormattedParameter("--repetitions {0}")
             self.attach_info = FormattedParameter("--path {0}")
+            self.use_daos_agent_env = FormattedParameter("--use-daos-agent-env {0}")
```
Comment (sylviachanoiyee) [inline src/tests/ftest/network/cart_self_test.py:49]: are we using dahs or underscore?

File: src/tests/ftest/util/apricot/apricot/test.py [hunk 12]
```diff
@@ -727,6 +727,7 @@ def write_string_to_logfile(self, message):
             cart_ctl.cfg_path.value = "."
             cart_ctl.m.value = message
             cart_ctl.n.value = None
+            cart_ctl.use_daos_agent_env.value = "--use_daos_agent_env"
```
Comment (sylviachanoiyee) [inline src/tests/ftest/util/apricot/apricot/test.py:730]: are we using dash or underscore?

File: src/tests/ftest/util/cart_ctl_utils.py [hunk 13]
```diff
@@ -63,6 +63,7 @@ def __init__(self, namespace, command):
         self.l = FormattedParameter("-l {}")
         self.n = BasicParameter("-n")
         self.m = FormattedParameter("-m {}")
+        self.use_daos_agent_env = BasicParameter("--use_daos_agent_env")
```
Comment (sylviachanoiyee) [inline src/tests/ftest/util/cart_ctl_utils.py:66]: dash or underscore?

File: src/utils/ctl/cart_ctl.c [hunk 14]
```diff
@@ -304,6 +312,9 @@ parse_args(int argc, char **argv)
 	} else if (strcmp(argv[1], "add_log_msg") == 0) {
 		/* avoid checkpatch warning */
 		ctl_gdata.cg_cmd_code = CMD_LOG_ADD_MSG;
+	} else if (strcmp(argv[1], "use_daos_agent_env") == 0) {
```
Comment (sylviachanoiyee) [inline src/utils/ctl/cart_ctl.c:315]: dash or underscore?

File: src/utils/ctl/cart_ctl.c [hunk 15]
```diff
@@ -319,11 +330,12 @@ parse_args(int argc, char **argv)
 		{"log_mask", required_argument, 0, 'l'},
 		{"no_sync", optional_argument, 0, 'n'},
 		{"message", required_argument, 0, 'm'},
+		{"use_daos_agent_env", no_argument, 0, 'u'},
```
Comment (sylviachanoiyee) [inline src/utils/ctl/cart_ctl.c:333]: dahs or underscore?

File: src/utils/ctl/cart_ctl.c [hunk 16]
```diff
@@ -263,6 +267,8 @@ print_usage_msg(const char *msg)
 	printf("\tdon't perform 'wait for ranks' sync\n");
 	printf("-m 'log_message'\n");
 	printf("\tSpecify log message to be sent to remote server\n");
+	printf("--use-daos-agent-env\n");
```
Comment (sylviachanoiyee) [inline src/utils/ctl/cart_ctl.c:270]: dash or underscore?

File: src/tests/ftest/network/cart_self_test.yaml [hunk 17]
```diff
@@ -16,6 +16,7 @@ self_test:
   repetitions: 1000
   endpoint: 0:0
   max_inflight_rpcs: 16
+  use_daos_agent_env: ""
```
Comment (sylviachanoiyee) [inline src/tests/ftest/network/cart_self_test.yaml:19]: Is this suppose to overwrite the default in apricot/test.py?

File: src/tests/ftest/network/cart_self_test.py (hunk 9 above)
Comment (sylviachanoiyee) [reply src/tests/ftest/network/cart_self_test.py:48]: My only concern here is if we dump attachinfo and then point selftest to it, how do we know it is daos agent aware?

File: src/tests/ftest/util/apricot/apricot/test.py (hunk 12 above)
Comment (sylviachanoiyee) [reply src/tests/ftest/util/apricot/apricot/test.py:730]: I think it'll be good to keep consistency at least for --use_daos_agent_env in this pr.

File: src/cart/SConscript [hunk 18]
```diff
@@ -91,11 +81,8 @@ def scons():
 
     denv = env.Clone(LIBS=[])
 
-    # Generate cart utility shared objects
-    build_utility_shared_obj(denv)
-
     denv.AppendUnique(LIBS=['gurt'])
-    prereqs.require(denv, 'mercury')
+    prereqs.require(denv, 'mercury', 'protobufc')
```
Comment (jolivier23) [inline src/cart/SConscript:85]: is protobuf needed here?

File: src/utils/ctl/cart_ctl.c (hunk 15 above)
Comment (mallove79) [reply src/utils/ctl/cart_ctl.c:335]: [Underscore](https://github.com/daos-stack/daos/commit/1d72e32378e9f66349d0b133118ffeee5d6d6212#diff-3cdaae73b3be35c105f461ab5a5476fe7ff5511617f2dc4786c25fa3a8c25646R270).

File: src/utils/ctl/cart_ctl.c (hunk 16 above)
Comment (mallove79) [reply src/utils/ctl/cart_ctl.c]: [Underscore](https://github.com/daos-stack/daos/commit/1d72e32378e9f66349d0b133118ffeee5d6d6212#diff-3cdaae73b3be35c105f461ab5a5476fe7ff5511617f2dc4786c25fa3a8c25646R270).

File: src/cart/crt_utils.c [hunk 19]
```diff
@@ -372,31 +490,44 @@ crtu_cli_start_basic(char *local_group_name, char *srv_group_name,
 	rc = pthread_create(progress_thread, NULL, crtu_progress_fn, crt_ctx);
 	D_ASSERTF(rc == 0, "pthread_create() failed; rc=%d\n", rc);
 
-	if (use_cfg) {
-		while (attach_retries-- > 0) {
-			rc = crt_group_attach(srv_group_name, grp);
-			if (rc == 0)
+	if (! use_daos_agent_env) {
```
[the remark this answers is not in the document]
Comment (mallove79) [reply src/cart/crt_utils.c]: a7a392ebeb634c4bd428fa52186e852442752056

File: src/cart/SConscript (hunk 18 above)
Comment (mallove79) [reply src/cart/SConscript]: `protobufc` isn't needed there.  It's only needed in [src/cart/utils](https://github.com/daos-stack/daos/pull/6219/files/74092f55cc5e7ac5e09bfd83f0d7f80c39926787#diff-29b3cefb1e87c5dce6289b7576dee25213ca370580650564b5a7f70e1ed0e661R32).

**Review by jolivier23 — APPROVED** [review]

File: src/tests/ftest/util/cart_ctl_utils.py (hunk 13 above)
Comment (phender) [reply src/tests/ftest/util/cart_ctl_utils.py:66]: Assuming `--use_daos_agent_env` is an optional command argument w/o a value (similar to something like `--debug` or, for the `daos_agent` command, `--insecure`) then it would be better implemented as:
```suggestion
        self.use_daos_agent_env = FormattedParameter("--use_daos_agent_env", False)
```
This definition makes `--use_daos_agent_env` a fixed entity and the `self.use_daos_agent_env.value` is then set to either `True` or `False` to determine if it is included as part of the command string.  Using `BasicParameter` opens us up to the possibility of someone redefining the `--use_daos_agent_env` string.

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/util/apricot/apricot/test.py (hunk 12 above)
Comment (phender) [reply src/tests/ftest/util/apricot/apricot/test.py:744]: With the change to using `use_daos_agent_env = FormattedParameter("--use_daos_agent_env", False)` the `.value` should be set to `True` to enable using the command line argument (https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils_base.py#L160-L161):
```suggestion
            cart_ctl.use_daos_agent_env.value = True
```
Sorry for not mentioning this explicitly in my previous review comment.

**Review by phender — DISMISSED** [review]

File: src/utils/ctl/cart_ctl.c (hunk 14 above)
Comment (mallove79) [reply src/utils/ctl/cart_ctl.c:317]: [Underscore](https://github.com/daos-stack/daos/commit/1d72e32378e9f66349d0b133118ffeee5d6d6212#diff-3cdaae73b3be35c105f461ab5a5476fe7ff5511617f2dc4786c25fa3a8c25646R270).

File: src/tests/ftest/network/cart_self_test.py (hunk 10 above)
Comment (mallove79) [reply src/tests/ftest/network/cart_self_test.py:86]: https://github.com/daos-stack/daos/commit/d4022c644a8d1f51b48ec3f97ff0ec9cdadee0d8#diff-ca093b7ae7587c88bc15cd3bd5d64269766f939a55c829b79ee6ec6c1d8c02c3L89

File: src/cart/utils/crt_utils.c [hunk 20]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
 
+#define SYS_INFO_BUF_SIZE 16
+
+int
+crtu_dc_mgmt_net_cfg_rank_add(const char *name, crt_group_t *group,
+		    crt_context_t *context)
+{
+	int	i;
+	int rc = 0;
+	struct dc_mgmt_sys_info crt_net_cfg_info;
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:360]: visually align variables

File: src/cart/utils/crt_utils.c [hunk 21]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
+	Mgmt__GetAttachInfoResp *crt_net_cfg_resp;
+
+	/* Query the agent for the CaRT network configuration parameters */
+	rc = get_attach_info(name, true /* all_ranks */, &crt_net_cfg_info, &crt_net_cfg_resp);
+	D_ASSERTF(rc == 0, "get_attach_info failed, rc=%d\n", rc);
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:365]: we have opts.assert_on_error (see line 47) that each test sets. we should only assert if its true

File: src/cart/utils/crt_utils.c [hunk 22]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
+crtu_dc_mgmt_net_cfg_setenv(const char *name)
+{
+	int rc;
+	char buf[SYS_INFO_BUF_SIZE];
+	char *crt_timeout;
+	char *ofi_interface;
+	char *ofi_domain;
+	char *cli_srx_set;
+	struct dc_mgmt_sys_info crt_net_cfg_info;
+	Mgmt__GetAttachInfoResp *crt_net_cfg_resp;
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:402]: visually align vars

File: src/cart/utils/crt_utils.c [hunk 23]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
+	if (rc != 0)
+		D_GOTO(cleanup, rc = d_errno2der(errno));
+
+	/* If the server has set this, the client must use the same value. */
+	if (crt_net_cfg_info.srv_srx_set != -1) {
+		sprintf(buf, "%d", crt_net_cfg_info.srv_srx_set);
+		rc = setenv("FI_OFI_RXM_USE_SRX", buf, 1);
+		if (rc != 0)
+			D_GOTO(cleanup, rc = d_errno2der(errno));
+		D_INFO("Using server's value for FI_OFI_RXM_USE_SRX: %s\n",
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:424]: should be probably d_debug

File: src/cart/utils/crt_utils.c [hunk 24]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
+
+	/* Allow client env overrides for these three */
+	crt_timeout = getenv("CRT_TIMEOUT");
+	if (!crt_timeout) {
+		sprintf(buf, "%d", crt_net_cfg_info.crt_timeout);
+		rc = setenv("CRT_TIMEOUT", buf, 1);
+		if (rc != 0)
+			D_GOTO(cleanup, rc = d_errno2der(errno));
+	} else {
+		D_INFO("Using client provided CRT_TIMEOUT: %s\n",
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:444]: d_debug

File: src/cart/utils/crt_utils.c [hunk 25]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
+			crt_timeout);
+	}
+
+	ofi_interface = getenv("OFI_INTERFACE");
+	if (!ofi_interface) {
+		rc = setenv("OFI_INTERFACE", crt_net_cfg_info.interface, 1);
+		if (rc != 0)
+			D_GOTO(cleanup, rc = d_errno2der(errno));
+	} else {
+		D_INFO("Using client provided OFI_INTERFACE: %s\n",
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:454]: d_debug

File: src/cart/utils/crt_utils.c [hunk 26]
```diff
@@ -375,31 +507,41 @@ crtu_cli_start_basic(char *local_group_name, char *srv_group_name,
...
-
-		/* load group info from a config file and
-		 * delete file upon return
-		 */
-		rc = crtu_load_group_from_file(grp_cfg_file, *crt_ctx, *grp,
-					       -1, true);
-		D_ASSERTF(rc == 0, "crtu_load_group_from_file() failed;"
-			  "rc=%d\n", rc);
+		crtu_dc_mgmt_net_cfg_setenv(srv_group_name);
+		crtu_dc_mgmt_net_cfg_rank_add(srv_group_name, *grp, *crt_ctx);
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:544]: should check rc and exit on errors

File: src/client/api/event.c [hunk 27]
```diff
@@ -59,6 +59,12 @@ static unsigned int eq_ref;
  */
 static tse_sched_t daos_sched_g;
 
+void
+daos_eq_ref_set(int value)
+{
+    eq_ref = value;
+}
```
Comment (frostedcmos) [inline src/client/api/event.c:66]: what purpose does this serve? why do we have to set eq_ref value at all?

File: src/cart/utils/crt_utils.c [hunk 28]
```diff
@@ -346,12 +349,141 @@ crtu_load_group_from_file(const char *grp_cfg_file, crt_context_t ctx,
...
+				DP_RC(rc));
+			goto err_group;
+		}
+	}
+
+err_group:
+	if (rc != 0) {
+		crt_group_view_destroy(group);
+	}
+
```
Comment (frostedcmos) [inline src/cart/utils/crt_utils.c:388]: since you are not preserving attach info across calls, should this not do put attac info at the end?

File: src/mgmt/cli_mgmt.c [hunk 29]
```diff
@@ -197,7 +197,7 @@ free_get_attach_info_resp(Mgmt__GetAttachInfoResp *resp)
 	mgmt__get_attach_info_resp__free_unpacked(resp, &alloc.alloc);
 }
 
-static void
+void
```
Comment (frostedcmos) [inline src/mgmt/cli_mgmt.c:200]: instead of this, you probably want to create wrapper functions such as dc_put_attach_info() dc_get_attach_info() that would be publically exposed and call static functions from those wrappers

**Review by frostedcmos — CHANGES_REQUESTED** [review]

File: src/client/api/event.c (hunk 27 above)
Comment (mallove79) [reply src/client/api/event.c:66]: Without it, we fail on this assertion:
```
self_test: src/client/api/event.c:163: daos_get_crt_ctx: Assertion `eq_ref > 0' failed.
```
Originally I used `daos_eq_lib_init()`, but that seems to have unwanted side effects (e.g., calls `crt_init()`, creates contexts, etc.)

File: src/cart/utils/crt_utils.c (hunk 21 above)
Comment (mallove79) [reply src/cart/utils/crt_utils.c:365]: Ah, I see.  So this option allows us to force tests to proceed despite an error?  We currently set `assert_on_error` to `true` for all tests in [tests/ftest/cart](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/cart).  I assume we want the `opts.assert_on_error` condition for all `D_ASSERTF`'s, so I'll add it to other places in `crt_utils.c` where it's missing.

File: src/cart/utils/crt_utils.c (hunk 28 above)
Comment (mallove79) [reply src/cart/utils/crt_utils.c:398]: When I try preserving attach info across calls, I hit this error:
```
[stderr] [1,0]<stderr>:self_test: src/mgmt/svc.pb-c.c:469: mgmt__get_attach_info_resp__free_unpacked: Assertion `message->base.descriptor == &mgmt__get_attach_info_resp__descriptor' failed.
```

**Review by frostedcmos — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]
